ocaml/xapi/test_vm.ml

```diff
@@ -0,0 +1,82 @@
+(*
+ * Copyright (C) 2013 Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open OUnit
+open Test_common
+open Xapi_vm
+open QCheck
+
+(*--- Xapi_vm.set_bios_strings ---*)
+
+let test_vm_set_bios_strings () =
+  let __context = make_test_database () in
+  let self = make_vm ~__context () in
+
+  (* Test empty value *)
+  assert_raises_api_error
+    Api_errors.invalid_value
+    ~args:["enclosure-asset-tag"; "Value provided is empty"]
+    (fun () -> Xapi_vm.set_bios_strings ~__context ~self ~value:[`enclosureassettag, ""]);
+
+  (* Test value having more than 512 charactors *)
+  let value = String.make (Xapi_globs.bios_string_limit_size + 1) 'x' in
+  let err = Printf.sprintf "%s has length more than %d characters" value Xapi_globs.bios_string_limit_size in
+  assert_raises_api_error
+    Api_errors.invalid_value
+    ~args:["enclosure-asset-tag"; err]
+    (fun () -> Xapi_vm.set_bios_strings ~__context ~self ~value:[`enclosureassettag, value]);
+
+  (* Test non printable ascii values *)
+  let ascii1 = Gen.int_bound 31 in
+  let ascii2 = Gen.int_range 127 255 in
+  let ascii_range = (Gen.generate 10 ascii1) @ (Gen.generate 10 ascii2) in
+  let non_printable_strs = List.map (fun c -> Printf.sprintf "%c" (Char.chr c)) ascii_range in
+  List.iter (fun str ->
+    assert_raises_api_error Api_errors.invalid_value
+    ~args:["enclosure-asset-tag"; str ^ " has non-printable ASCII characters"]
+    (fun () -> Xapi_vm.set_bios_strings ~__context ~self ~value:[`enclosureassettag, str])
+  ) non_printable_strs;
+
+  (* Test setting bios_strings is reflected in VM record *)
+  let random_str () =
+    let str_gen = printable_string_of_size (Gen.int_range 1 512) in
+    List.fold_left (fun acc str -> acc ^ str) "" (Stdext.Xstringext.String.split '\n' (Gen.generate1 str_gen.gen))
+  in
+  let test_map =
+    [`biosvendor, random_str ();
+     `biosversion, random_str ();
+     `systemmanufacturer, random_str ();
+     `systemproductname, random_str ();
+     `systemversion, random_str ();
+     `systemserialnumber, random_str ();
+     `enclosureassettag, random_str ()]
+  in
+  Xapi_vm.set_bios_strings ~__context ~self ~value:test_map;
+  let custom_strings = List.map (fun (k, v) -> (Record_util.vm_bios_key_to_string k, v)) test_map in
+  let bios_strings =
```

Comment:
This is essentially a copy of the function from xapi_vm.ml. I think that is the wrong approach to unit testing. What we should do here is actually hardcode the input and expected output, and check whether the function returns the right thing. Or even better, a number of different inputs and expected outputs. The functors from `Test_highlevel` make this easy; see test_cpuid_helpers.ml for examples. I also don't think that it is useful to use random strings.

Reply:
Since the result of `set_bios_strings` can be an exception, you can use the Either module as done here: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/test_cpuid_helpers.ml#L395. Using this approach, you only need one instance of the test functor, and one list of test cases that cover everything including the exceptions.

Reply:
Thats a neat way, Thanks @robhoes for the suggestion. Please have a look now.